src/pybind/rbd/rbd.pyx

```diff
@@ -371,10 +371,10 @@ ELSE:
     cdef rados_ioctx_t convert_ioctx(rados.Ioctx ioctx) except? NULL:
         return <rados_ioctx_t>ioctx.io
 
-cdef int progress_callback(uint64_t offset, uint64_t total, void* ptr) with gil:
+cdef int progress_callback(uint64_t offset, uint64_t total, void* ptr) except? -1 with gil:
```

Comment:
The "pybind/rbd: be explicit about exceptions in callback definitions" title seems misleading to me.  AFAIU Cython 3 changed the default behaviour of non-extern `cdef` functions to propagate exceptions by default.  By adding `except? -1` and `except *` you are making a change for Cython 2 where, instead of swallowing exceptions in `on_progress` and `oncomplete` callbacks, we are propagating them, right?  This kind of change is definitely needs to be described better and probably worth a release note.

Reply:
> you are making a change for Cython 2 where, instead of swallowing exceptions in `on_progress` and `oncomplete` callbacks, we are propagating them

If this is correct, let's add a test to `src/test/pybind/test_rbd.py` where `progress_cb()` raises an exception and we assert that it's propagated.  I'm thinking of something similar to `test_remove_canceled()` but with e.g. `raise EOFError` instead of `return -ECANCELED` in the body.  And then a similar test for the completion callback.

Reply:
Actually my understanding was that adding `except? -1` and `except *` changed nothing for Cython 2, as it was the default behavior here and I just explicitly tells it now. And for Cython 3 the behavior will be different from its default behavior. Though my understanding may be wrong. Anyway, I am going to add the callback tests as you suggested and check the behavior with cython2 before and after my change.

Reply:
I was going by https://cython.readthedocs.io/en/0.29.x/src/userguide/language_basics.html which seems to advise to tag on `noexcept` to keep the existing behavior:
>To explicitly mark a function as not returning an exception use noexcept.
>
>    cdef int spam() noexcept:
>            ...
>
>This is worth doing because (a) “explicit is better than implicit”, and (b) the default behaviour for cdef functions will change in Cython 3.0 so that functions will propagate exceptions by default. Therefore, it is best to mark them now if you want them to swallow exceptions in the future.

Reply:
Adding `noexcept` keyword was the first thing I tried. Unfortunately cython < 3 knows nothing about it and fails to compile, so I would need something like `IF CYTHON_VERSION >= ...` preprocessing to have different definitions for different versions, but I failed to find a simple solution.

Reply:
> Adding `noexcept` keyword was the first thing I tried.

I think we should agree on the desired behavior first: swallow exceptions (as Cython 2 does, presumably) or propagate them (as Cython 3 does, again presumably -- pending your confirmation with the help of those tests).  I think propagating exceptions is better, but if they are currently swallowed and this wasn't raised as an issue for so many years (not to my knowledge, at least), maybe it doesn't matter that much?

Reply:
> I think propagating exceptions is better

... so I like the PR as is, but having tests is always good (and needed to confirm behavior between Cython2 and Cython3 here anyway).